src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -78,20 +78,13 @@ private void StopRaisingEvents()
                 {
                     // Stop the RunLoop and wait for the thread to exit gracefully
                     Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
-                    _watcherThread.Join();
-                    _watcherRunLoop = IntPtr.Zero;
                 }
 
-                // Clean up the EventStream, if it exists
-                if (!_eventStream.IsInvalid)
-                {
-                    _eventStream = new SafeEventStreamHandle(IntPtr.Zero);
-                }             
-
-                // Cleanup the callback
-                if (_callback != null)
+                // Gurd against us from calling Join on our own background thread (for 
+                // example, calling Dispose or Stop from a callback).
+                if (Thread.CurrentThread != _watcherThread)
                 {
-                    _callback = null;
+                    _watcherThread.Join();
```

Comment:
Each Start call recreates all information necessary (the Thread, EventSource, RunLoop, etc) so that shouldn't be a problem. The only (potential) thing I can see is a race condition that should be next to impossible to hit and would be a no-op if it; this condition is the rapid start-stop-start scenario. The cleaning up thread would have to be delayed so long that the start thread would recreate all of it's information, send a Sync command to the Disk, and make all those interop calls before the cleanup thread made it to the next two commands; since this is **extremely unlikely** to ever happen outside of debugging, and if it did happen it would be a no-op, it should be safe.